src/components/app/index.tsx

```diff
@@ -2,56 +2,176 @@ import { h, Component } from 'preact';
 import { bind, bitmapToImageData } from '../../lib/util';
 import * as style from './style.scss';
 import Output from '../output';
+import Options from '../options';
 
-import {MozJpegEncoder} from '../../lib/codec-wrappers/mozjpeg-enc';
+import { Encoder } from '../../codecs/codec';
+import IdentityEncoder from '../../codecs/identity/encoder';
+import MozJpegEncoder from '../../codecs/mozjpeg/encoder.worker';
+
+export type ImageType = 'original' | 'MozJpeg';
+
+export type CodecOptions = any;
+
+const ENCODER_NAMES = {
+  original: 'Original Image',
+  MozJpeg: 'JPEG'
+};
+
+const ENCODERS = {
+  original: IdentityEncoder,
+  MozJpeg: MozJpegEncoder
+};
+
+type Image = {
+  type: ImageType,
+  options: CodecOptions,
+  data?: ImageBitmap,
+  counter: number,
+  loading: boolean
+};
 
 type Props = {};
 
 type State = {
-  img?: ImageBitmap
+  sourceFile?: File,
+  sourceImg?: ImageBitmap,
+  sourceData?: ImageData,
+  images: Array<Image>,
+  loading: boolean,
```

Comment:
Since this is _the_ app state, can we add some comments explaining the props? Like, what does `loading` stand for and what is the desired effect?